Move the SystemFallbackCache to FontCache

mattwoodrow opened the pull request:
#### 38edf7c001ddfea00c90e7db40ca57b0a02c2ebd
<pre>
Move the SystemFallbackCache to FontCache
<a href="https://bugs.webkit.org/show_bug.cgi?id=233972">https://bugs.webkit.org/show_bug.cgi?id=233972</a>

Reviewed by Myles C. Maxfield.

With OffscreenCanvas, the SystemFallbackCache can be touched from
worker threads. This cache contains AtomStrings, which are thread
specific.

Move the SystemFallbackCache code from inside FontCache.cpp to separate
files (and rename it to SystemFallbackFontCache), to avoid FontCache.h
from getting too noisy. Also take the opportunity to wrap it in a class
to give it meaningful methods.

* platform/graphics/FontCache.h:
(WebCore::FontCache::systemFontFallbackCache):
* platform/graphics/Font.cpp:
* platform/graphics/FontCache.h:
(WebCore::FontCache::systemFontFallbackCache):
* platform/graphics/Font.cpp:
(WebCore::systemFallbackCache):
* platform/graphics/SystemFallbackFontCache.cpp:
(WebCore::SystemFallbackFontCache::forCurrentThread):
Move SystemFallbackCache from a singleton to FontCache.

* platform/graphics/Font.h:
(WebCore::Font::setIsUsedInSystemFallbackFontCache):
(WebCore::Font::isUsedInSystemFallbackFontCache const):
New functions to note a Font is in a SystemFallbackFontCache, since the
method that adds it to and checks its presence in the cache is no longer
in Font.

* platform/graphics/Font.cpp:
(WebCore::Font::~Font):
Call a new SystemFallbackFontCache::forCurrentThread function.

* platform/graphics/Font.cpp:
(WebCore::Font::systemFallbackFontForCharacter const):
(WebCore::Font::removeFromSystemFallbackCache):
* platform/graphics/SystemFallbackFontCache.cpp: Added.
(WebCore::SystemFallbackFontCache::systemFallbackFontForCharacter):
(WebCore::SystemFallbackFontCache::remove):
Move lookup and removal methods from Font to SystemFallbackFontCache.

* Headers.cmake:
* Sources.txt:
* WebCore.xcodeproj/project.pbxproj:
* platform/graphics/Font.cpp:
(): Deleted.
(WebCore::operator==): Deleted.
(WebCore::CharacterFallbackMapKeyHash::hash): Deleted.
(WebCore::CharacterFallbackMapKeyHash::equal): Deleted.
(WebCore::CharacterFallbackMapKeyHashTraits::constructDeletedValue): Deleted.
(WebCore::CharacterFallbackMapKeyHashTraits::isDeletedValue): Deleted.
* platform/graphics/SystemFallbackFontCache.h: Added.
(WebCore::SystemFallbackFontCache::CharacterFallbackMapKey::operator== const):
(WebCore::SystemFallbackFontCache::CharacterFallbackMapKeyHash::hash):
(WebCore::SystemFallbackFontCache::CharacterFallbackMapKeyHash::equal):
(WebCore::SystemFallbackFontCache::CharacterFallbackMapKeyHashTraits::constructDeletedValue):
(WebCore::SystemFallbackFontCache::CharacterFallbackMapKeyHashTraits::isDeletedValue):
Move existing SystemFallbackFontCache code from Font.cpp to a new file and
wrap it in a class.

* platform/graphics/freetype/FontCacheFreeType.cpp:
(WebCore::systemFallbackFontSetCache):
(WebCore::FontCache::systemFallbackForCharacters):
(WebCore::FontCache::platformPurgeInactiveFontData):
(WebCore::systemFallbackCache):
Rename the SystemFallbackCache in FontCacheFreeType.cpp, which is used
for a different purpose from the one that lived in Font.cpp.

Canonical link: <a href="https://commits.webkit.org/252618@main">https://commits.webkit.org/252618@main</a>
</pre>

cdumez commented on Source/WebCore/page/cocoa/MemoryReleaseCocoa.mm.orig:
```diff
@@ -0,0 +1,137 @@
+#import "config.h"
+#import "MemoryRelease.h"
+
+<<<<<<< HEAD
```
Yeah, that looks bad :)

Also, was this already reviewed? The PR was just created but the change log already says "Reviewed by Myles"

mattwoodrow replied:
Oh dear :(

Yeah it was reviewed in bugzilla a while ago, trying to get things landed.

mcatanzaro commented on Source/WebCore/platform/graphics/freetype/FontCacheFreeType.cpp:
```diff
@@ -239,16 +239,17 @@ struct FallbackFontDescriptionKeyHash {
+static SystemFallbackFontSetCache& systemFallbackFontSetCache()
 {
-    static NeverDestroyed<SystemFallbackCache> cache;
+    // FIXME: This should be moved to FontCache since it can be accessed from worker threads.
```
Hm, I think _none_ of the font code is threadsafe. @carlosgcampos, do you have time to look into this?

heycam committed:
Move the SystemFallbackCache to FontCache
https://bugs.webkit.org/show_bug.cgi?id=233972

Reviewed by Myles C. Maxfield.

With OffscreenCanvas, the SystemFallbackCache can be touched from
worker threads. This cache contains AtomStrings, which are thread
specific.

Move the SystemFallbackCache code from inside FontCache.cpp to separate
files (and rename it to SystemFallbackFontCache), to avoid FontCache.h
from getting too noisy. Also take the opportunity to wrap it in a class
to give it meaningful methods.

* platform/graphics/FontCache.h:
(WebCore::FontCache::systemFontFallbackCache):
* platform/graphics/Font.cpp:
* platform/graphics/FontCache.h:
(WebCore::FontCache::systemFontFallbackCache):
* platform/graphics/Font.cpp:
(WebCore::systemFallbackCache):
* platform/graphics/SystemFallbackFontCache.cpp:
(WebCore::SystemFallbackFontCache::forCurrentThread):
Move SystemFallbackCache from a singleton to FontCache.

* platform/graphics/Font.h:
(WebCore::Font::setIsUsedInSystemFallbackFontCache):
(WebCore::Font::isUsedInSystemFallbackFontCache const):
New functions to note a Font is in a SystemFallbackFontCache, since the
method that adds it to and checks its presence in the cache is no longer
in Font.

* platform/graphics/Font.cpp:
(WebCore::Font::~Font):
Call a new SystemFallbackFontCache::forCurrentThread function.

* platform/graphics/Font.cpp:
(WebCore::Font::systemFallbackFontForCharacter const):
(WebCore::Font::removeFromSystemFallbackCache):
* platform/graphics/SystemFallbackFontCache.cpp: Added.
(WebCore::SystemFallbackFontCache::systemFallbackFontForCharacter):
(WebCore::SystemFallbackFontCache::remove):
Move lookup and removal methods from Font to SystemFallbackFontCache.

* Headers.cmake:
* Sources.txt:
* WebCore.xcodeproj/project.pbxproj:
* platform/graphics/Font.cpp:
(): Deleted.
(WebCore::operator==): Deleted.
(WebCore::CharacterFallbackMapKeyHash::hash): Deleted.
(WebCore::CharacterFallbackMapKeyHash::equal): Deleted.
(WebCore::CharacterFallbackMapKeyHashTraits::constructDeletedValue): Deleted.
(WebCore::CharacterFallbackMapKeyHashTraits::isDeletedValue): Deleted.
* platform/graphics/SystemFallbackFontCache.h: Added.
(WebCore::SystemFallbackFontCache::CharacterFallbackMapKey::operator== const):
(WebCore::SystemFallbackFontCache::CharacterFallbackMapKeyHash::hash):
(WebCore::SystemFallbackFontCache::CharacterFallbackMapKeyHash::equal):
(WebCore::SystemFallbackFontCache::CharacterFallbackMapKeyHashTraits::constructDeletedValue):
(WebCore::SystemFallbackFontCache::CharacterFallbackMapKeyHashTraits::isDeletedValue):
Move existing SystemFallbackFontCache code from Font.cpp to a new file and
wrap it in a class.

* platform/graphics/freetype/FontCacheFreeType.cpp:
(WebCore::systemFallbackFontSetCache):
(WebCore::FontCache::systemFallbackForCharacters):
(WebCore::FontCache::platformPurgeInactiveFontData):
(WebCore::systemFallbackCache):
Rename the SystemFallbackCache in FontCacheFreeType.cpp, which is used
for a different purpose from the one that lived in Font.cpp.

Canonical link: https://commits.webkit.org/252618@main
